src/test/net_tests.cpp

```diff
@@ -771,4 +773,129 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
 }
 
+std::vector<NodeEvictionCandidate> GetRandomNodeEvictionCandidates(const int n_candidates, FastRandomContext& random_context)
+{
+    std::vector<NodeEvictionCandidate> candidates;
+    for (int i = 0; i < n_candidates; ++i) {
+        candidates.push_back({/* id */ static_cast<NodeId>(random_context.randrange(100)),
+                              /* nTimeConnected */ static_cast<int64_t>(random_context.randrange(100)),
+                              /* nMinPingUsecTime */ static_cast<int64_t>(random_context.randrange(100)),
+                              /* nLastBlockTime */ static_cast<int64_t>(random_context.randrange(100)),
+                              /* nLastTXTime */ static_cast<int64_t>(random_context.randrange(100)),
+                              /* fRelevantServices */ random_context.randbool(),
+                              /* fRelayTxes */ random_context.randbool(),
+                              /* fBloomFilter */ random_context.randbool(),
+                              /* nKeyedNetGroup */ random_context.randrange(100),
+                              /* prefer_evict */ random_context.randbool(),
+                              /* m_is_local */ random_context.randbool()});
+    }
+    return candidates;
+}
+
+bool IsEvicted(const std::vector<NodeEvictionCandidate>& candidates, const std::vector<NodeId> node_ids, FastRandomContext& random_context)
+{
+    std::vector<NodeEvictionCandidate> shuffled_candidates = candidates;
+    Shuffle(shuffled_candidates.begin(), shuffled_candidates.end(), random_context);
+    const Optional<NodeEvictionCandidate> evicted_node = SelectNodeToEvict(std::move(shuffled_candidates));
+    if (!evicted_node) {
+        return false;
+    }
+    return std::find(node_ids.begin(), node_ids.end(), evicted_node->id) != node_ids.end();
+}
+
+bool IsEvicted(const int number_of_nodes, std::function<void(NodeEvictionCandidate&, const size_t)> candidate_setup_fn, const std::vector<NodeId> node_ids, FastRandomContext& random_context)
+{
+    std::vector<NodeEvictionCandidate> candidates = GetRandomNodeEvictionCandidates(number_of_nodes, random_context);
+    for (size_t i = 0; i < candidates.size(); ++i) {
+        candidate_setup_fn(candidates[i], i);
+    }
+    return IsEvicted(candidates, node_ids, random_context);
+}
+
+namespace {
+// The number of eviction candidates at (or above) which we are guaranteed that
+// the current eviction logic will find a node to evict no matter what eviction
+// candidates it is given. This constant may need to be adjusted if the eviction
+// logic changes.
+const int GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE = 29;
+
+// The number of eviction candidates at (or below) which we are guaranteed that
+// the current eviction logic won't find a node to evict no matter what eviction
+// candidates it is given. This constant may need to be adjusted if the eviction
+// logic changes.
+const int GUARANTEED_NON_EVICTION_AT_N_CANDIDATES_OR_BELOW = 20;
```

Comment:
Yes, but I'm afraid I cannot help you with that at the moment: so far it is a purely an empirical result :D

Counterexamples or proofs welcome! :)

Reply:
Possible reason: the protection at [net.cpp::961](https://github.com/bitcoin/bitcoin/blob/fabecce/src/net.cpp#L961) for up to 8 non-tx-relay peers may, or may not apply to the randomly generated eviction candidates since `!n.fRelayTxes && n.fRelevantServices` will evaluate to randbool. So we cannot count them in the _guaranteed_ eviction?